src/planner/partialize.c

```diff
@@ -185,6 +197,652 @@ partialize_agg_paths(RelOptInfo *rel)
+		return append_path->subpaths;
+	}
+	else if (IsA(path, MergeAppendPath))
+	{
+		MergeAppendPath *merge_append_path = castNode(MergeAppendPath, path);
+		return merge_append_path->subpaths;
+	}
+	else if (ts_is_chunk_append_path(path))
+	{
+		CustomPath *custom_path = castNode(CustomPath, path);
+		return custom_path->custom_paths;
+	}
+	else if (handle_gather_path && IsA(path, GatherPath))
+	{
+		return get_subpaths_from_append_path(castNode(GatherPath, path)->subpath, false);
+	}
+
+	/* Aggregation push-down is not supported for other path types so far */
+	return NIL;
+}
+
+/*
+ * Copy an AppendPath and set new subpaths.
+ */
+static AppendPath *
+copy_append_path(AppendPath *path, List *subpaths)
+{
+	AppendPath *newPath = palloc(sizeof(AppendPath));
+	memcpy(newPath, path, sizeof(AppendPath));
+	newPath->subpaths = subpaths;
+	cost_append(newPath);
+
+	return newPath;
+}
+
+/*
+ * Copy a MergeAppendPath and set new subpaths.
+ */
+static MergeAppendPath *
+copy_merge_append_path(PlannerInfo *root, MergeAppendPath *path, List *subpaths)
+{
+	MergeAppendPath *newPath = create_merge_append_path_compat(root,
+															   path->path.parent,
+															   subpaths,
+															   path->path.pathkeys,
+															   NULL,
+															   path->partitioned_rels);
+
+#if PG14_LT
+	newPath->partitioned_rels = list_copy(path->partitioned_rels);
+#endif
+
+	newPath->path.param_info = path->path.param_info;
+
+	return newPath;
+}
+
+/*
+ * Copy an append like path and set new subpaths
+ */
+static Path *
+copy_append_like_path(PlannerInfo *root, Path *path, List *new_subpaths)
+{
+	if (IsA(path, AppendPath))
+	{
+		AppendPath *append_path = castNode(AppendPath, path);
+		return (Path *) copy_append_path(append_path, new_subpaths);
+	}
+	else if (IsA(path, MergeAppendPath))
+	{
+		MergeAppendPath *merge_append_path = castNode(MergeAppendPath, path);
+		return (Path *) copy_merge_append_path(root, merge_append_path, new_subpaths);
+	}
+	else if (ts_is_chunk_append_path(path))
+	{
+		CustomPath *custom_path = castNode(CustomPath, path);
+		ChunkAppendPath *chunk_append_path = (ChunkAppendPath *) custom_path;
+		return (Path *) ts_chunk_append_path_copy(chunk_append_path, new_subpaths);
+	}
+
+	/* Should never happen, already checked by caller */
+	Ensure(false, "Unknown path type");
+	pg_unreachable();
+}
+
+/*
+ * Generate a partially sorted aggregated agg path on top of a path
+ */
+static Path *
+create_sorted_partially_agg_path(PlannerInfo *root, Path *path, PathTarget *target,
+								 double d_num_groups, GroupPathExtraData *extra_data)
+{
+	Query *parse = root->parse;
+
+	/* Determine costs for aggregations */
+	AggClauseCosts *agg_partial_costs = &extra_data->agg_partial_costs;
+
+	int presorted_keys;
+	bool is_sorted =
+		pathkeys_count_contained_in(root->group_pathkeys, path->pathkeys, &presorted_keys);
+
+	if (!is_sorted)
+	{
+		path = (Path *) create_sort_path(root, path->parent, path, root->group_pathkeys, -1.0);
+	}
```

Comment:
This code is close to the planner code of PostgreSQL; the planner code of PostgreSQL uses [shadowing a lot](https://github.com/postgres/postgres/blob/04a09ee944ac904c7fb7ca8f7230a524f514b3e7/src/backend/optimizer/plan/planner.c#L6872). 

Changing this might make it harder to adapt to future changes. But I have no strong opinion here.